### PR TITLE
Make Loadout Optimizer 25x faster

### PR DESCRIPTION
**Pull request by bhollis**
I've been beating my head against the wall forever trying to find a way to algorithmically improve Loadout Optimizer, with no luck. However, it turns out that just unrolling a lot of the loops manually, and avoiding object creation, makes huge improvements to runtime speed. I'm embarrassed I didn't fully go down this road earlier, but some minor attempts I'd made weren't promising and I didn't realize how far I'd have to go. I also sorta hoped V8 would do a better job of automatically unrolling loops but it doesn't (or we weren't triggering it).

Beyond the loop unrolling and removing object allocation, I also switched the statMix insertion code from insertion sort to binary search, which was a noticeable performance win, and cleaned up a few other areas that contributed to some speedups. Embarrassingly, we were also checking whether to report progress far too often, and reducing how often that runs was another big speedup.

On my machine (2013-era 2.6 GHz Intel Core i7, Chrome), processing goes from ~250,000 combos/sec to ~6,900,000 combos/sec. I can process my entire messy vault of 300+ armor items in less than 5 seconds. That's less than 400 CPU clock cycles per set, which is getting close to ideal - perhaps with WASM SIMD we could get better but not likely by another order of magnitude.

This speed boost is so significant that I've removed all the code related to reducing armor combinations to a reasonable size - we now always chew through every single combination, exhaustively. This means not only are you now guaranteed to get all the best sets, but the "Max tier" display on every stat is also accurate, so I'm enabling that on prod.

As one other positive side effect we no longer cap stats at 0/100 for display, so the tooltips over each set will show the actual value.

I have a proof of concept for a followup PR that introduces concurrent processing of sets on multiple threads with a simple partition of the problem that seems to cap out at 15,000,000 combos/second on my machine, which would be 60x faster than the original code.

### CODE REVIEW

File: src/app/loadout-builder/process-worker/process.ts [hunk 1]
```diff
@@ -412,25 +259,22 @@ export function process(
...
+            let tiersString = '';
+            for (let index = 0; index < 6; index++) {
+              const value = Math.min(Math.max(stats[index], 0), 100);
+              const tier = tiers[index];
               // Make each stat exactly one code unit so the string compares correctly
-              const filter = statFilters[statHash];
+              const filter = statFiltersInStatOrder[index];
               if (!filter.ignored) {
-                tiers += tier.toString(11);
+                tiersString += tier.toString(16);
```
Comment (bhollis) [inline src/app/loadout-builder/process-worker/process.ts:273]: Computers like powers of 2 - this was a noticeable speedup

Comment (nev-r) [reply src/app/loadout-builder/process-worker/process.ts:273]: we live in hell

Comment (ryan-rushton) [reply src/app/loadout-builder/process-worker/process.ts:273]: Lets stick a comment on this so no one changes it

File: src/app/loadout-builder/process-worker/process.ts [hunk 2]
```diff
@@ -280,110 +126,107 @@ export function process(
...
-                  statRangeExceeded = true;
-                }
-                totalTier += tier;
+            for (let index = 0; index < 6; index++) {
+              const tier = tiers[index];
+              const filter = statFiltersInStatOrder[index];
+              if (!filter.ignored && (tier > filter.max || tier < filter.min)) {
+                statRangeExceeded = true;
               }
             }
```
Comment (ryan-rushton) [inline src/app/loadout-builder/process-worker/process.ts:231]: Would unrolling this loop make any difference?

File: src/app/loadout-builder/process-worker/process.ts [hunk 3]
```diff
@@ -398,7 +241,11 @@ export function process(
               continue;
             }
 
+            const armor = [helm, gaunt, chest, leg, classItem];
```
Comment (ryan-rushton) [inline src/app/loadout-builder/process-worker/process.ts:244]: Another optimisation could be to put off creating this array until we know we are going to keep the set.

**Comment by nev-r** [comment]
given what look like the same settings,
imageimage

beta is giving me about 40 options for t26 loadouts, and dev seems to give me 1.
has the sorting changed or something??
image

**Comment by nev-r** [comment]
and beta doesn't even find that specific t26 (5-3-8-10).
none of its t26s include dragon's shadow at all.
image

it skips from 5-7 to 4-9

**Comment by delphiactual** [comment]
it seems the mapping between the selection pane and the display pane are wrong
changing min recovery is affecting min mobility
image

**Comment by bhollis** [comment]
Ah, I must not be mapping to stat order correctly somewhere

File: src/app/loadout-builder/process-worker/process.ts (hunk 3 above)
Comment (bhollis) [reply src/app/loadout-builder/process-worker/process.ts:244]: That's what this is - it's needed right away in `canTakeSlotIndependentMods`

**Review by nev-r — APPROVED** [review]
results looking comparable to beta now and taking 1/5 of the time

File: src/app/loadout-builder/process-worker/process.ts (hunk 2 above)
Comment (bhollis) [reply src/app/loadout-builder/process-worker/process.ts]: It could, but it's relatively difficult to do so.

**Pull request merged**